### PR TITLE
MINIFICPP-1975 Volatile and persistent combination of repositories should be avoided

### PR DESCRIPTION
**Pull request by lordgamez**
https://issues.apache.org/jira/browse/MINIFICPP-1975

---------
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: controller/Controller.h [hunk 1]
```diff
@@ -259,13 +261,16 @@ std::shared_ptr<org::apache::nifi::minifi::core::controller::ControllerService>
...
   }
-
   flow_repo->initialize(configuration);
 
-
   configuration->get(org::apache::nifi::minifi::Configure::nifi_content_repository_class_name, content_repo_class);
 
   const std::shared_ptr content_repo = org::apache::nifi::minifi::core::createContentRepository(content_repo_class, true, "content");
+  const bool is_flow_repo_non_persistent = flow_repo->isNoop() || std::dynamic_pointer_cast<org::apache::nifi::minifi::core::repository::VolatileFlowFileRepository>(flow_repo) != nullptr;
+  const bool is_content_repo_non_persistent = std::dynamic_pointer_cast<org::apache::nifi::minifi::core::repository::VolatileContentRepository>(content_repo) != nullptr;
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply controller/Controller.h]: Good point, updated in 99e3282df555128cd885a57a5ffacfcc6ff15478

[the remark this answers is not in the document]
Comment (adamdebreceni) [reply controller/Controller.h]: does `typeid` handle inheritance as well? if we create a `FasterVolatileContentRepo` I presume `typeid` based comparison will fail, as we do it once I do not think speed is of concern, being volatile (or persistent) seems to be a property of any repository, maybe we should let the objects themselves report their capabilities through a virtual method?

[the remark this answers is not in the document]
Comment (lordgamez) [reply controller/Controller.h]: I suppose it there is also an `isNoop` function for repositories an `isVolatile` could also be introduced as well. @szaszm what do you think?

**Review by fgerlits — APPROVED** [review]

File: controller/Controller.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (szaszm) [reply controller/Controller.h]: `typeid` doesn't check for inheritance. If you think this is an important requirement, we can keep the dynamic_cast or introduce an `isVolatile` member. I'm fine with both approaches, although keeping the dynamic_cast sounds simpler.

**Review by szaszm — APPROVED** [review]

File: controller/Controller.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (lordgamez) [reply controller/Controller.h:269]: Reverted to using dynamic_cast

**Review by adamdebreceni — APPROVED** [review]

**Pull request closed without merging**